pkg/generator/generate_cert.go

```diff
@@ -167,146 +179,194 @@ func (rc *GenerateCert) generateCA(ctx context.Context, CASecretName string, nam
-	}
+		// Read the node certificate into memory
+		pemCert, err := ioutil.ReadFile(filepath.Join(rc.CertsDir, "node.crt"))
+		if err != nil {
+			return errors.Wrap(err, "unable to read node.crt")
+		}
 
-	// Read the node certificate into memory
-	pemCert, err := ioutil.ReadFile(filepath.Join(rc.CertsDir, "node.crt"))
-	if err != nil {
-		return errors.Wrap(err, "unable to read node.crt")
-	}
+		validFrom, validUpto, err := rc.getCertLife(pemCert)
+		if err != nil {
+			return err
+		}
 
-	validFrom, validUpto, err := rc.getCertLife(pemCert)
-	if err != nil {
-		return err
-	}
+		// Read the node key into memory
+		pemKey, err := ioutil.ReadFile(filepath.Join(rc.CertsDir, "node.key"))
+		if err != nil {
+			return errors.Wrap(err, "unable to ready node.key")
+		}
 
-	// Read the node key into memory
-	pemKey, err := ioutil.ReadFile(filepath.Join(rc.CertsDir, "node.key"))
-	if err != nil {
-		return errors.Wrap(err, "unable to ready node.key")
+		// add certificate info in the secret annotations
+		annotations := resource.GetSecretAnnotations(validFrom, validUpto, rc.NodeCertConfig.Duration.String())
+
+		// create and save the TLS certificates into a secret
+		secret = resource.CreateTLSSecret(nodeSecretName, corev1.SecretTypeTLS,
+			resource.NewKubeResource(ctx, rc.client, namespace, kube.DefaultPersister))
+
+		if err = secret.UpdateTLSSecret(pemCert, pemKey, ca, annotations); err != nil {
+			return errors.Wrap(err, "failed to update node TLS secret certs")
+		}
+
+		logrus.Infof("Generated and saved node key and certificate in secret [%s]", nodeSecretName)
+
+		return nil
 	}
+	// check if the existing secret is ready to be consumed. If found ready, skip cert generation
+	if secret.Ready() && secret.ValidateAnnotations() {
 
-	// add certificate info in the secret annotations
-	annotations := resource.GetSecretAnnotations(validFrom, validUpto, rc.NodeCertConfig.Duration.String())
+		if rc.RotateNodeCert {
+			isRequired, reason := secret.IsRotationRequired(rc.NodeCertConfig.Duration, rc.NodeAndClientCronSchedule)
+			if isRequired {
+				logrus.Infof("Node Certificate: %s", reason)
 
-	// create and save the TLS certificates into a secret
-	secret = resource.CreateTLSSecret(nodeSecretName, corev1.SecretTypeTLS,
-		resource.NewKubeResource(ctx, rc.client, namespace, kube.DefaultPersister))
+				if err = generate(); err != nil {
+					return err
+				}
 
-	if err = secret.UpdateTLSSecret(pemCert, pemKey, ca, annotations); err != nil {
-		return errors.Wrap(err, "failed to update node TLS secret certs")
+				if err = kube.RollingUpdate(ctx, rc.client, rc.DiscoveryServiceName, namespace, rc.ReadinessWait); err != nil {
+					return
+				}
+				return nil
+			}
+		}
+
+		logrus.Infof("Node secret [%s] is found in ready state, skipping Node cert generation", nodeSecretName)
+		return nil
 	}
 
-	logrus.Infof("Generated and saved node key and certificate in secret [%s]", nodeSecretName)
+	return generate()
```

Comment:
Why are we doing the closure here? And also above? Can't we just inline the function?

Reply:
We are not manipulating any values in the function so kept it as closure. Changed it to inline function for better readability.